buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java

```diff
@@ -0,0 +1,85 @@
+ * the License.
+ */
+package io.netty.buffer.api.tests;
+
+import io.netty.buffer.api.Buffer;
+import io.netty.buffer.api.BufferAllocator;
+import io.netty.util.AsciiString;
+import org.checkerframework.checker.units.qual.A;
+import org.junit.Assert;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.nio.charset.StandardCharsets;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+public class BufferCharSequenceTest extends BufferTestSupport {
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    void readCharSequence(Fixture fixture) {
+        try (BufferAllocator allocator = fixture.createAllocator()) {
+            try (Buffer buf = allocator.allocate(32)) {
+                String data = "Hello World";
+                buf.writeBytes(data.getBytes());
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(0, buf.readerOffset());
+
+                final CharSequence charSequence = buf.readCharSequence(data.length(), US_ASCII);
+                Assertions.assertEquals(data, charSequence.toString());
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(data.length(), buf.readerOffset());
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    void writeCharSequence(Fixture fixture) {
+        try (BufferAllocator allocator = fixture.createAllocator()) {
+            try (Buffer buf = allocator.allocate(32)) {
+                AsciiString data = new AsciiString("Hello world".getBytes(US_ASCII));
+                buf.writeCharSequence(data, US_ASCII);
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(0, buf.readerOffset());
+
+                final byte[] read = readByteArray(buf);
+                Assertions.assertEquals(data.toString(), new String(read, US_ASCII));
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(data.length(), buf.readerOffset());
+            }
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    void readAndWriteCharSequence(Fixture fixture) {
+        try (BufferAllocator allocator = fixture.createAllocator()) {
+            try (Buffer buf = allocator.allocate(32)) {
+                AsciiString data = new AsciiString("Hello world".getBytes(US_ASCII));
+                buf.writeCharSequence(data, US_ASCII);
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(0, buf.readerOffset());
+
+                final CharSequence read = buf.readCharSequence(data.length(), US_ASCII);
+                Assertions.assertEquals(data, read);
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(data.length(), buf.readerOffset());
+            }
+        }
+    }
```

Comment:
We could also parameterise over a number of different encodings and `CharSequence` implementations.

Reply:
In the interest of time and the fact that we are depending on the contract of `CharSequence`, lemme punt that for now